src/util/util_log.h

```diff
@@ -1,7 +1,70 @@
 #pragma once
 
 #include "../nvapi_private.h"
+#include "util_string.h"
 
 namespace dxvk::log {
-    void write(const std::string& message);
+    namespace fmt {
+        inline std::string hnd(const void* h) {
+            if (h == nullptr)
+                return "hnd=0x0";
+
+            return str::format("hnd=", std::setw(8), std::hex, h);
+        }
+
+        inline std::string ptr(const void* p) {
+            if (p == nullptr)
+                return "nullptr";
+
+            return str::format("ptr=", std::setw(8), std::hex, p);
+        }
+
+        inline std::string flt(const float f) {
+            if (f == 0)
+                return "0.0";
+
+            if (f == 1)
+                return "1.0";
+
+            return str::format(f);
+        }
+
+        inline std::string flags(const uint32_t h) {
+            if (h == 0)
+                return "0x0";
+
+            return str::format("flags=", std::setw(4), std::hex, h);
+        }
+
+        inline std::string d3d12_cpu_descriptor_handle(D3D12_CPU_DESCRIPTOR_HANDLE h) {
+            return str::format("{ptr=", std::hex, h.ptr, "}");
+        }
+    }
+
+    bool tracing();
+
+    void write(const std::string& severity, const std::string& message);
+
+    inline void info(const std::string& message) {
+        log::write("info", message);
+    }
+
+    inline void trace1(std::stringstream&) {}
+
+    template <typename T, typename... Tx>
+    void trace1(std::stringstream& str, const T& arg, const Tx&... args) {
+        str << arg << ", ";
+        trace1(str, args...);
+    }
+
+    template <typename... Args>
+    void trace(const std::string& name, const Args&... args) {
+        std::stringstream stream;
+        trace1(stream, args...);
+        auto str = stream.str();
+        if (str.ends_with(", "))
+            str.erase(str.length() - 2);
```

Comment:
Maybe instead of chopping the resulting string just add an overload like so?

```cpp
template <typename T>
inline void trace1(std::stringstream& str, const T& arg)
{
    str << arg;
}
```

This way the trailing `, ` shouldn't be added in the first place and these two lines could be removed. Or at least I think that's how C++ works which I could be wrong about :sweat_smile:

Plus, even though those helper methods are only used by `trace`, I think naming them something like `append` instead of `trace1` would be more appropriate.

Reply:
Will try, but templating can be magic, so no guaranties ;).
Regarding the naming, I went with the style in `util_strings` which we (me) took over initially from dxvk. Agree that there is room for naming improvements ;).

Reply:
This worked, thanks!